plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -553,6 +581,76 @@ public ConnectorInsertTableHandle beginInsert(ConnectorSession session, Connecto
                 .collect(toImmutableList())));
     }
 
+    private void cleanExtraOutputFiles(ConnectorSession session, Set<String> writtenFiles)
+    {
+        HdfsContext hdfsContext = new HdfsContext(session);
+        Set<String> locations = getOutputFilesLocations(writtenFiles);
+        for (String location : locations) {
+            cleanExtraOutputFiles(hdfsContext, session.getQueryId(), location, writtenFiles);
+        }
+    }
+
+    private void cleanExtraOutputFiles(HdfsContext hdfsContext, String queryId, String location, Set<String> filesToKeep)
+    {
+        List<String> filesToDelete = new LinkedList<>();
+        try {
+            log.debug("Deleting failed attempt files from %s for query %s", location, queryId);
+            FileSystem fileSystem = hdfsEnvironment.getFileSystem(hdfsContext, new Path(location));
+            if (!fileSystem.exists(new Path(location))) {
+                // directory may nat exit if no files were actually written
+                return;
+            }
+
+            // files are written flat in a single directory so we do not need to list recursively
```

Comment:
I think the more appropriate reason is that we're looking for siblings of files we wrote.
thus, even if files are eg nested per partition, the code would still work.

Reply:
Yeah - but if they were nested withing given partition it would not. So the fact that we are writing directly into `location` not in a subdirectory of `location` is important here. 

Rephrase to `// files within given partition are written flat into location; we need to list recursively`